### ISSUE TITLE
Improve the format of http.Request.__repr__ to include eg class and object identity

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6391]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6391|
|Type|enhancement|
|Created|2013-03-26 13:09:39Z|
|Branch|https://github.com/twisted/twisted/tree/http-request-repr-6391|

In [#6120](https://github.com/twisted/twisted/issues/6120) radix commented that Request.__repr__ should include the
ClassName and memory address.

```
        19:24 < radix> rwall: hmm
        19:25 < radix> rwall: can you please include the class name and instance ID in that repr as well
        19:25 < radix> I really hate it when classes __repr__ don't include the basic info
        19:26 < radix> if you want me to put a comment on the ticket I can do that
        19:26 < radix> oh, you're reviewing, not authoring.
        19:26 < rwall> radix: Yes. That sounds like a good idea. Is it ok to change the repr without warning?
        19:27 < radix> I see. and it's not even changing the contents
        19:27 < rwall> Yep, just adding a missing test.
        19:27 < radix> rwall: yes, as tomprince said in an earlier comment on the ticket, we don't want users relying on the content of the __repr__
        19:27 < radix> but I see the content isn't even being changed in this branch, so it's probably not the appropriate branch to change it in.
        19:29 < rwall> radix: Ok, but I'll raise another ticket. What's the ideal repr? What should be the standard?
        19:30 < rwall> radix: and without wanting to dwell too long on such a small ticket...do you think __repr__ should have full docstring?
        19:32 < radix> huh, standards, that's an idea. we don't have any
        19:32 < radix> rwall: I don't think __repr__'s docstring needs to document the format absolutely. but I don't really think it matters either way.
        19:33 < rwall> ok
        19:34 < radix> Personally, off the cuff, I would suggest something like <ClassName at 0xffffffff ...> or something, where "..." could be "GET /" or                maybe even "method=GET path=/" etc
```

Here's an interesting discussion about ideal repr output.
 * http://stackoverflow.com/questions/452300/python-object-repr-self-should-be-an-expression

   If at all possible, this (repr) should look like a valid Python
   expression that could be used to recreate an object with the same
   value (given an appropriate environment).

   If this is not possible, a string of the form <...some useful
   description...> should be returned.

   The return value must be a string object. If a class defines repr()
   but not str(), then repr() is also used when an “informal” string
   representation of instances of that class is required.

   This is typically used for debugging, so it is important that the
   representation is information-rich and unambiguous

The current repr outputs method='GET' etc but in [#6120](https://github.com/twisted/twisted/issues/6120), exarkun
commented that he'd rather just see method=GET. I think his reasoning
was that ultimately, the method and clientproto should be supplied
and stored as twisted.python.constants instead of byte strings.  And
similarly, the URI should also probably be supplied and stored as a
URI object of some sort. This would enforce the use of byte strings
for these attribute values. (see related ticket:6168).

So here's a proposed format...

```
 <twisted.web.http.Request at 0xffffffff method=GET uri=http://www.example.com/foo/bar clientproto=HTTP/1.1>
```

See:
 * [twisted/web/http.py](twisted/web/http.py)

Attachments:

* [h.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1d2/1d29771d9cecc114fc00b25c11cd952a5585d535/a862530f67dd8cbf53078e295fed0ebc425f5bf6.patch) (440 bytes) - added by introom on 2013-03-27 12:01:27Z - modified Request.__repr__ function
* [6391_improved.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1d2/1d29771d9cecc114fc00b25c11cd952a5585d535/ec13df5d1e606837fa8a091ff37d80d366e3a36d.patch) (2408 bytes) - added by introom on 2013-03-28 13:42:09Z - added unit test
* [6391_improve2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/1d2/1d29771d9cecc114fc00b25c11cd952a5585d535/36a249652b9c9144276f63d1d92b39ea878439eb.patch) (2305 bytes) - added by introom on 2013-03-28 14:05:23Z - I modified the indention.

<details><summary>Searchable metadata</summary>

```
trac-id__6391 6391
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_http_request_repr_6391 branches/http-request-repr-6391
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1364303379000000 1364303379000000
changetime__1365260835000000 1365260835000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="introom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|introom commented|
|-|-|

I modified the Request.__repr__ function.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [37939]) Branching to 'http-request-repr-6391'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [37940]) Apple 6391_improve2.patch 

refs [#6391](https://github.com/twisted/twisted/issues/6391)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `new`|
|-|-|

Thanks!  A couple minor points:

  1. in twisted/web/test/test_http.py
    1. Lines should be wrapped at column 79.
    1. instead of `C{string}`, you should write `L{str}`.
    * L is for "links"... sort of.  Any time you have a name for an object that probably has documentation somewhere else, L makes sense.  If you just have some stuff that is more or less Python code (for example, "a + b"), C is appropriate.  
    * `str` is the name of a builtin type, _string_ is either the name of a stdlib module (which doesn't seem to be related to the docstring you're writing) or just a plain old english word.  Either way, you want `str` instead.
    1. The unit test should really be two unit tests, one for each form the repr result can take.  A good rule of thumb is one assertion per test.
    1. As a matter of style (so you can take this part of the review or leave it), the indentation convention where you have either:
    ```
assertEqual(foo,
            bar)
``` or
    ```
assertEqual(
    foo,
    bar)
``` seems preferable to the one where you have:
    ```
assertEqual(foo,
    bar)
```
  1. twisted/web/http.py
    1. in the docstring for `__repr__`, it doesn't seem necessary to duplicate the explanation of the format of the result.
    1. Same string/str, C/L comment as above
    1. I slightly disagree with the ticket description's request for "twisted.web.http.Request" to appear in the repr output, instead of just "Request".  The former is quite a mouthful, and you probably already know you're dealing with the web once you see the rest of the string.
    1. There's one extra feature not being tested here, which is that a subclass of a different name will get a slightly different repr result (reflecting the subclass name instead of just "Request").
  1. Needs a topfile, probably a misc.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

Well, oops, forgot to preview that comment and spend 45 minutes fixing the markup.  Sorry about that.  Anyhow, I went ahead and addressed those points myself - r37941.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/http-request-repr-6391)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince set owner to @exarkun|
|-|-|

I don't think including the exact format of `__repr__` in warranted. Users shouldn't be depending on the format. Please merge, after deleting that bit from the docstring.

Note: This supersedes [#6120](https://github.com/twisted/twisted/issues/6120), so that ticket should be closed with this one, as well.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [37976]) Vague out the `__repr__` docstring

refs [#6391](https://github.com/twisted/twisted/issues/6391)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [37977]) Merge http-request-repr-6391

Author: introom, exarkun
Reviewer: exarkun, tom.prince
Fixes: [#6391](https://github.com/twisted/twisted/issues/6391), [#6120](https://github.com/twisted/twisted/issues/6120)

Expand the output of `twisted.web.http.Request.__repr__` to include a bit
more information.  Also add unit tests for it.
